File: D_Recommendations.py
import sys
from math import inf

def read_int():
    return int(input())

def read_pair():
    a, b = map(int, input().split())
    return (a, b)

def ms(l, r, arr, res):
    if l > r:
        return
    if l == r:
        return

    m = (l + r) // 2

    ms(l, m, arr, res)
    ms(m + 1, r, arr, res)

    left = arr[l:m + 1]
    right = arr[m + 1:r + 1]

    merged = []
    li, ri = 0, 0

    while li < len(left) and ri < len(right):
        if left[li][1] >= right[ri][1]:
            merged.append(right[ri])
            res[right[ri][2]] = min(res[right[ri][2]], left[li][1])
            ri += 1
        else:
            merged.append(left[li])
            li += 1

    while li < len(left):
        merged.append(left[li])
        li += 1

    while ri < len(right):
        merged.append(right[ri])
        ri += 1

    for i in range(len(merged)):
        arr[l + i] = merged[i]

def main():
    t = read_int()
    for _ in range(t):
        n = read_int()
        sngs = []
        sngsc = []
        cnt = {}

        for i in range(n):
            p = read_pair()
            sngs.append((p[0], p[1], i))
            sngsc.append((p[0], p[1], i))
            cnt[(p[0], p[1])] = cnt.get((p[0], p[1]), 0) + 1

        lres = [inf] * n
        rres = [inf] * n

        sngs.sort(key=lambda x: (x[0], -x[1]))

        ms(0, n - 1, sngs, rres)

        for i in range(len(sngs)):
            sngs[i] = (-sngs[i][1], -sngs[i][0], sngs[i][2])

        sngs.sort(key=lambda x: (x[0], -x[1]))

        ms(0, n - 1, sngs, lres)

        for i in range(n):
            if cnt[(sngsc[i][0], sngsc[i][1])] >= 2:
                print(0)
                continue

            lf = -lres[i]
            rf = rres[i]
            if lf == -inf or rf == inf:
                print(0)
            else:
                print(sngsc[i][0] - lf + rf - sngsc[i][1])

if __name__ == "__main__":
    main()

File: D_Recommendations.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
#include <map>
#include <tuple>

using namespace std;

#define inf INT_MAX
#define ii() read_int()
#define li() read_pair()

typedef tuple<int, int, int> Song;

int read_int() {
    int x;
    cin >> x;
    return x;
}

pair<int, int> read_pair() {
    int a, b;
    cin >> a >> b;
    return {a, b};
}

void ms(int l, int r, vector<Song>& arr, vector<int>& res) {
    if (l > r) return;
    if (l == r) return;

    int m = (l + r) / 2;

    ms(l, m, arr, res);
    ms(m + 1, r, arr, res);

    vector<Song> left(arr.begin() + l, arr.begin() + m + 1);
    vector<Song> right(arr.begin() + m + 1, arr.begin() + r + 1);

    vector<Song> merged;
    size_t li = 0, ri = 0;

    while (li < left.size() && ri < right.size()) {
        if (get<1>(left[li]) >= get<1>(right[ri])) {
            merged.push_back(right[ri]);
            res[get<2>(right[ri])] = min(res[get<2>(right[ri])], get<1>(left[li]));
            ri++;
        } else {
            merged.push_back(left[li]);
            li++;
        }
    }

    while (li < left.size()) {
        merged.push_back(left[li]);
        li++;
    }

    while (ri < right.size()) {
        merged.push_back(right[ri]);
        ri++;
    }

    for (size_t i = 0; i < merged.size(); i++) {
        arr[l + i] = merged[i];
    }
}

int main() {
    int t = ii();
    while (t--) {
        int n = ii();
        vector<Song> sngs;
        vector<Song> sngsc;
        map<pair<int, int>, int> cnt;

        for (int i = 0; i < n; i++) {
            auto p = li();
            sngs.push_back({p.first, p.second, i});
            sngsc.push_back(sngs.back());
            cnt[{p.first, p.second}]++;
        }

        vector<int> lres(n, inf), rres(n, inf);

        sort(sngs.begin(), sngs.end(), [](const Song& a, const Song& b) {
            return make_pair(get<0>(a), -get<1>(a)) < make_pair(get<0>(b), -get<1>(b));
        });

        ms(0, n - 1, sngs, rres);

        for (auto& song : sngs) {
            song = {-get<1>(song), -get<0>(song), get<2>(song)};
        }

        sort(sngs.begin(), sngs.end(), [](const Song& a, const Song& b) {
            return make_pair(get<0>(a), -get<1>(a)) < make_pair(get<0>(b), -get<1>(b));
        });

        ms(0, n - 1, sngs, lres);

        for (int i = 0; i < n; i++) {
            if (cnt[{get<0>(sngsc[i]), get<1>(sngsc[i])}] >= 2) {
                cout << 0 << endl;
                continue;
            }

            int lf = -lres[i], rf = rres[i];
            if (lf == -inf || rf == inf) {
                cout << 0 << endl;
            } else {
                cout << get<0>(sngsc[i]) - lf + rf - get<1>(sngsc[i]) << endl;
            }
        }
    }

    return 0;
}
